Update to compatible with aiida-core 2.x

unkcpz opened the pull request:
Hi @greschd @bosonie,

Here are some attempts to remove the dependency of `aiida-pytest`. All test for `run` passed but not for `submit` caused by issue I opened at https://github.com/aiidateam/aiida-core/issues/5511. Do you have better suggestion on how to do that?

Then, since we drop 1.x, other thing I can image to apply here is that we can use dot `.` to access and construct namespace instead of double score `__` for `nested_links`. Whatelse any new features you think we can update and implemented for this tool?

I can also manage to update the pre-commit dependencies and release workflow if you prefer, I can open another PR for that.

unkcpz committed:
Drop aiida-pytest dependent

unkcpz committed:
remove depentencies of aiida-pytest replace with fixture from aiida-core

unkcpz committed:
drop aiida 1.x dependency

greschd commented on setup.json:
```diff
@@ -29,7 +29,7 @@
   "reentry_register": true,
   "install_requires": [
-    "aiida-core>=1.3.0<2.0.0",
+    "aiida-core>=2.0.0",
```
Because we must assume that `3.x` will again be backwards-incompatible (otherwise it probably wouldn't be a major release), the version constraint here should be `>=2.0.0<3.0.0`.

greschd reviewed (commented):
> Then, since we drop 1.x, other thing I can image to apply here is that we can use dot . to access and construct namespace instead of double score __ for nested_links.

Yeah, I think that'd make sense. Also agree on updating the pre-commit hooks etc.

As for the `submit` tests: I'm not entirely sure what's different between this code and what I was doing in `aiida-pytest`; maybe it's the `AIIDA_PATH` environment variable that needs to be set?

https://github.com/greschd/aiida_pytest/blob/823d8066fd37288132e31314de511c2cf991538d/aiida_pytest/_configure.py#L107

As for the necessity of running the test with both `run` and `submit` at all.. I'm not entirely sure if this is still needed. There used to be subtle differences where some things couldn't get serialized properly on `submit`, but this is just as much a test of `aiida-core` functionality as it tests `aiida-optimize`.
TL;DR: If it's simple enough to keep the `submit` tests around I'm all for it, but if that ends up being the major blocker I would be open to removing them.

greschd commented on setup.json:
```diff
@@ -29,7 +29,7 @@
   ],
   "reentry_register": true,
```
Could it be that this is no longer needed?

unkcpz replied:
Indeed!

unkcpz commented:
> TL;DR: If it's simple enough to keep the submit tests around I'm all for it, but if that ends up being the major blocker I would be open to removing them.

@greschd thanks! I'll wait a bit too see whether it is easy for https://github.com/aiidateam/aiida-core/issues/5511 to be fixed. Then we just need to add `submit` option back to fixture. 

EDIT: could you approve the running of CI test? thanks.

greschd commented:
The ` --quiet-wipe --print-status` are flags from `aiida-pytest`, those should be removed in the pipeline.

unkcpz committed:
Drop submit tests

unkcpz committed:
Add daemon submit as nightly daemon CI test

unkcpz committed:
test_daemon to launch daemon and submit process

unkcpz committed:
pre-commit fixes

unkcpz commented:
I borrow the concept from `aiida-core`, start a daemon and submit the process there for testing. But only running one submit test there for testing purposes. If you are good with this change I will add at least one test for every engine and wrapper.

greschd commented on .github/workflows/ci.yml:
```diff
@@ -20,7 +20,7 @@ jobs:
       - name: Install python dependencies
         env:
           INSTALL_TYPE: docs
```
I think the `INSTALL_TYPE` is now unused, right?

greschd commented:
> I borrow the concept from `aiida-core`, start a daemon and submit the process there for testing. But only running one submit test there for testing purposes. If you are good with this change I will add at least one test for every engine and wrapper.

Yep, makes sense :+1:

bosonie commented:
Hi there! What is the status of this? 
Tests on python 3.7 must be removed since aiida 2.0.0 does not support it anymore.
Please let me know if I can help somehow. Otherwise a release after this merge will be appreciated, so I can subsequently release the `aiida-siesta` version compatible with 2.0. Thanks!

unkcpz commented:
Hi @bosonie, I think this PR is almost ready. I didn't get time to manage to add daemon tests for every engine last week, but I think it is okay to add them in a separate PR? Meanwhile, I want to leave cleanup double underscore `__` into a separate PR, since I make some attempts and I think using `__` to access the node as it stores is more clear for now. 
I'll remove `INSTALL_TYPE` as @greschd mentioned and we can try to merge this.

unkcpz committed:
remove INSTALL_TYPE and drop all py3.7

greschd commented on .github/system_tests/sample_processes.py:
```diff
@@ -0,0 +1 @@
+../../tests/sample_processes.py
```
Uhmm.. how does this work?

unkcpz replied:
It is simply a soft link to the `sample_processes.py` 😆. Not sure it is clear, but otherwise have to copy and paste code from there.

greschd replied:
Ah yeah that's fine, I guess I was puzzled by the rendering online.

greschd commented on .github/system_tests/test_daemon.py:
```diff
@@ -0,0 +1,137 @@
+    """Check if jobs with given pks have finished."""
+    finished_list = [load_node(pk).is_terminated for pk in pks]
+    node_list = [load_node(pk) for pk in pks]
+    num_finished = len([_ for _ in finished_list if _])
```
Nitpick: `sum(finished_list)`?

unkcpz replied:
I personally don't quite a fan of sum a bool list. But fine to change if you prefer.

greschd replied:
Nah, it's fine to leave as-is.

greschd commented on .github/system_tests/test_daemon.py:
```diff
@@ -0,0 +1,137 @@
+        if not node.is_terminated:
+            print(f'not terminated: {node.pk} [{node.process_state}]')
+    print(f'{num_finished}/{len(finished_list)} finished')
+    return False not in finished_list
```
`return all(finished_list)` is clearer IMO

unkcpz replied:
The method is not used as you mentioned below.

greschd commented on .github/system_tests/test_daemon.py:
```diff
@@ -0,0 +1,137 @@
+        print(f'Note: the command failed, message: {exception}')
+
+
+def jobs_have_finished(pks):
```
Hmm wait.. is this function used anywhere?

unkcpz replied:
Yes, it is not used at last. Purged.

greschd commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,62 @@
+name: nightly
```
So these are _just_ the daemon tests, correct? I'm not sure they warrant a separate pipeline, or if they should be a step in the "regular" CI.

Since this is a much smaller project than `aiida-core`, I'd like to keep the testing infrastructure at a smaller level also.

unkcpz replied:
Fair point, I move it to ci.yml.

greschd commented on aiida_optimize/engines/_particle_swarm.py:
```diff
@@ -110,8 +110,8 @@ def submit_initialize(self):
         self.velocities = np.zeros((n_parts, n_vars))
-        for line in range(len(self.velocities)):
-            for col in range(len(self.velocities[line])):
+        for line in range(len(self.velocities)):  # pylint: disable=consider-using-enumerate
```
Hmm, I tend to agree with `pylint` here: `for line, _ in enumerate(self.velocities)` looks nicer to me. But both are fine, if you feel strongly about it.

unkcpz replied:
Agree.

greschd commented on tests/wrappers/test_add_inputs_wrapper.py:
```diff
@@ -2,18 +2,19 @@
 
 
-def test_basic(configure_with_daemon, echo_process):
+@pytest.mark.usefixtures('aiida_profile_clean')
```
Can we `autouse` this fixture?

unkcpz replied:
I think only tests that involve actually running the calculation need to clean the profile after finished.

greschd reviewed (commented):
Thanks for all your work @unkcpz!

Sorry for being a bit slow to review, I've been out sick a few days.

Anyways, I've got mostly some nitpicks / questions. In cases where you disagree, feel free to let me know I'm wrong :smile:

unkcpz committed:
review

- disable=consider-using-enumerate
- remove job_have_finished
- daemon test -> ci.yml fmt

unkcpz commented:
> Sorry for being a bit slow to review, I've been out sick a few days.
Anyways, I've got mostly some nitpicks / questions. In cases where you disagree, feel free to let me know I'm wrong 😄

No worries, I hope you are feeling better. I am also busy these days, so sorry for not being able to manage to add more submit tests and implement all things nicely. 

I update the PR, hope everything is clear.

greschd commented:
Aand merged! :tada:

Thanks again @unkcpz!
